rat/ast.py

```diff
@@ -582,6 +680,31 @@ def __str__(self):
         return f"InverseLogit({self.subexpr})"
 
 
+@dataclass
+class IfElse(Expr):
+    condition: Expr
+    true_expr: Expr
+    false_expr: Expr
+
+    def __iter__(self):
+        yield self
+        for expr in itertools.chain(self.condition, self.true_expr, self.false_expr):
+            yield expr
+
+    def __post_init__(self):
+        signatures = {
+            (types.BooleanType, types.IntegerType, types.IntegerType): types.IntegerType,
+            (types.BooleanType, types.RealType, types.RealType): types.RealType,
+            # (types.BooleanType, types.NumericType, types.NumericType): types.NumericType,
+            # this signature is not supported, because the backend, jax.lax.select, requires that the types of true
+            # and false be equal
```

Comment:
Yeah strong typing rocks.

Except we've got a problem with using a data in ifelse. Consider the following code, where `sigma` is from data:
```
y' ~ normal(theta[school], ifelse(school > 4, sigma, 1.0));
```

In `ifelse(school > 4, sigma, 1.0)`, since `sigma` is data, it's inferred to be a `Numeric` data type. And 1.0 is Real. This raises an error, since their types are different. Maybe we need a `float()` casting function for this scenario. (Float to int would be handled by floor, ceil, and round)

Reply:
Oh ouch. I think you're right with a float cast.

The other thing we could consider is strict types for the inputs. Is Numeric there just for the dataframe? If we always had a type would we need numeric?

Reply:
I'm getting a type error here:

```
mu[school] ~ normal(
      ifelse(
        school' == 0,
        10.0,
        mu[shift(school, 1)]
      ),
      0.1
    );
```

Invalid signature, boolean, real, numeric. Why is the parameter mu getting marked as numeric?

Reply:
Jeeez lol. This is literally the example you posted lolol. Well that happened quickly.

Reply:
Oh no wait, mu here is a parameter

Reply:
Hmm, but actually maybe this should be another error. I'm like using mu recursively here but it's being defined in the same place...

Reply:
Okay this fails:

```

    y ~ normal(mu[school], 10.0);
    mu[school]' ~ normal(
      ifelse(
        school == 0,
        10.0,
        mu[shift(school, 1)]
      ),
      0.1
    );
```

And I think it should pass

Reply:
Uh oh now getting `Input Type signature ['Boolean', 'Numeric', 'Numeric'] is invalid` time to investigate.

Reply:
Oh I see:

```
# (types.BooleanType, types.NumericType, types.NumericType): types.NumericType,
# this signature is not supported, because the backend, jax.lax.select, requires that the types of true
# and false be equal
```

float function seems like the quick escape. I will try to add.

Reply:
I think we should get rid of the Numeric type if ifelse can't support it. Presumably it's there for a reason though -- so I guess the question is, what do we need to do to get rid of Numeric or otherwise solve the conundrum?

Reply:
We can also just go be sloppy with types lol -- that's always an option we can just forget the first two comments here where we both proclaim the virtues of strong typing.

Reply:
Yeah numeric rn is just for data - I just pretended we don't know the type of data except it being numeric.

Reply:
I can see a simple fix is to give data a default type of real, and if the user knows that it's just pure integers, they can use int() to cast it to integers. Or vice versa.

Reply:
> what do we need to do to get rid of Numeric or otherwise solve the conundrum?

Only time real/int matters is for discrete distributions. I think it's safe to assume that the default type of parameters/data can be set as real, and give the user to downcast to int.

Reply:
Here's an interesting one:

```python
def test_ifelse_recursive_assignment():
    data_df = pandas.read_csv(os.path.join(test_dir, "eight_schools.csv"))

    data_df = pandas.DataFrame({
        "school" : [1, 2, 3, 4],
        "offset" : [10.0, 0.0, 0.0, 0.0]
    })

    model_string = """
    # This first line more or less shouldn't matter
    offset' ~ normal(mu[school], 1000.0);
    mu[school]' = ifelse(
        school == 1,
        10.0,
        mu[shift(school, 1)] + 1.0
    ) + epsilon;
    epsilon ~ normal(0.0, 0.1);
    """

    model = Model(data_df, model_string)
    fit = model.optimize()

    mu_df = fit.draws("mu")

    for row in mu_df.itertuples():
       assert row.mu == pytest.approx(9.0 + row.school, abs=1e-3, rel=1e-3)
```

The error seems to be that the `school == 1` expression isn't getting handled correctly in the recursive assignment (it's still a big vector thing instead of a scalar)

Reply:
Okay I resolved this for now with:

```
# This first line more or less shouldn't matter
offset' ~ normal(mu[school], 1000.0);
mu[school]' = ifelse(
  school[school] == 1,
  10.0,
  mu[shift(school, 1)] + 1.0
) + epsilon[school];
epsilon[school] ~ normal(0.0, 0.1);
```

Reply:
I'll make a separate issue for that. I added a float function (gonna rename it to real) but it looks like we can also work around this by doing stuff like `variable + 1.0` to get it to be real.

Reply:
It might be convenient to keep Numeric around just as an input type, but maybe we can make a rule of not returning any Numeric types?

Reply:
Yeah, numeric shouldn't exist as a return type.

Reply:
The school thing messing up in the condition is weird. `mu[school]` should be a length 8 vector, `data[school]` should also be a length 8 vector.

Reply:
Oh and my bad, parameters apparently were also assigned 'Numeric".

Reply:
Yeah I just made a separate issue for this (https://github.com/bbbales2/regressions/issues/85) -- seems like some edge case with the recursive assignment stuff. No need to fix here since there's workaround.

Reply:
I also added an issue for the types (https://github.com/bbbales2/regressions/issues/86). I think that can be pulled outta this thread too.

In which case, I'm happy with the tests so if they pass I'll merge (I'm assuming you're okay with the test changes I made)